README.md

## How to install HandyStuff

### To install from Github using R:

````r
install.packages("devtools")
library("devtools")
install_github(repo = "bryanhanson/HandyStuff@master")
library("HandyStuff")
````
If you use `@devel` you can get the development branch if it is available (and there may be other branches out there too).  Development branches have new, possibly incompletely tested features.  They may may also not be ready to pass checks and thus may not install automatically using the method above.  Check the news file to see what's up.

=======
Questions?  hanson@depauw.edu


NEWS


NEWS file for package HandyStuff
Bryan A. Hanson DePauw University, Greencastle Indiana USA
URL: github.com/bryanhanson/HandyStuff

Changes in version 0.6.157 2022-07-29
     + rxnNorm gains a new option for type = "raw"

Changes in version 0.6.130 2018-12-04
     + Built and checked against R 3.6 devel.
     + examplePage now writes to a tempdir, and the path defaults to getwd().

Changes in version 0.6.127 2017-08-08
     + Improved documentation for sumStatDeluxe.
     + Built and checked against R 3.5 RC.

Changes in version 0.6.123 2015-07-28
     + rxnNorm gains a legend argument to control the position of the legend.  This allows the data to be more prominent, since no space is wasted with a separate legend.

Changes in version 0.6.96 2015-07-21
     + Improved formatting of the aov table in rxnNorm using a new function, prettyAOVtable, and other tables were made to appear similarly.
     + All functions updated with if(!requireNamespace)... statements to coordinate with Suggests field.

Changes in version 0.6.56 2015-07-20
     + Brought into compliance with new CRAN namespace policies.
     + Overhauled rxnNorm as gridExtra had significant changes.

Changes in version 0.6.8 2015-05-06
     + Clean up a lot of small inconsistencies.
     + Overhauled compareCats, esp. the coloring of points.
     + Some changes to theme details.
     + The poster argument of compareCats and rxnNorm was replaced with a theme argument for greater flexibility.
     + lmEqn now has a theme argument too.
     + Updated package title.
     + Made some repairs to examplePage.  Now seems to work correctly.

Changes in version 0.6.7 2015-04-20
     + Added functions lpi and lpr.
     + Moved many Imports to Suggests.

Changes in version 0.6.6 2015-04-19
     + Spectroscopy related functions moved to package SpecHelpers
     + Package converted to roxygen2 documentation flow.
     + NEWS file format changed.
     + Many new functions added.  You'll just have to look for yourself.
     + compareCats seems to have a bug.  I'll fix it later.
     + Functions that had an "L" at the end of their name have lost it.
     + New style CITATION.

Changes in version 0.6-5 2014-01-14
     + Small document fixes.
     + Improvements to plotNMRspec.
     + Set exclude = NULL in examplePage.
     + I see a problem with colors of data points in compareCatsL.

Changes in version 0.6-4 2014-07-27
     + Significant repairs to screenThemeL and posterThemeL.
     + Fixed mis-information in rxnNormL.Rd

Changes in version 0.6-3 2013-06-10
     + Fixed a problem with column names in compareCatL.

Changes in version 0.6-2 2013-05-1
     + Fixed a problem with rounding in the plotNMRspec-related functions.

Changes in version 0.6-1 2013-04-13
     + Function compareCatsL which is the lattice equivalent of compareCats was added.
     + Function rxnNormL which is the lattice equivalent of rxnNorm was added.
     + Functions rxnNorm and compareCats which use ggplot2 were put on leave, awaiting repairs  so they work with the new ggplot2.
     + The following function arguments were changed to use a formula interface: compareCatsL, rxnNormL, lmEqn, lmEqnL.
     + Added function examplePage, slightly modified from the version by Sasha Epskamp at sachaepskamp.com/blog/HTMLexamples
     + Added functions posterThemeL and screenThemeL which control the appearance of the graphics output.  Currently these have the SAME settings.
     + Updates to documentation and github site.

Changes in version 0.5-6 2012-07-12
     + All functions have been restored and modified to work with the new ggplot2.  At least I think all is well! (tested with ggplot2_0.9.1)

Changes in version 0.5-5 2011-12-24
     + Numerous changes for 2.14 and with the current ggplot2!
     + Due to changes in ggplot2, ggExtra has been deprecated.  As a result the  table features in rxnNorm have been disabled until the functionality returns.  Sorry.  As a further result, gridExtra and ggExtra have been removed from NAMESPACE and DESCRIPTION.

Changes in version 0.5-4 2011-12-14
     + Modified plotNMRspec and makeSpec so that spectra are drawn consistently regardless of the x.range settings.  Now a consistent number of data points is used and peak height/area/appearance should not change when x.range is changed.

Changes in version 0.5-3 2011-10-31
     + Modified theme_poster and undo_theme_poster to give finer and more deliberate control over plotting parameters.  Should be much clearer if you have to hack the code.
     + Created a NAMESPACE and made other changes for compatibility with R 2.14.

Changes in version 0.5-2 2011-08-01
     + OK, this time I really added the ability to draw an ANOVA summary table to rxnNorm!

Changes in version 0.5-1 2011-7-29
     + Added the ability to draw an ANOVA summary table to rxnNorm.

Changes in version 0.5 2011-7-26
     + Modified gatherSpecFiles and gatherCsv so that a data frame is returned (This was claimed in the help file but not true).  So the data frame is written out as All Specs.csv and saved in the workspace.  Also, the LoggerPro export format .csv is now supported.

R/rxnNorm.R
#'
#' Create a Reaction Norm or Interaction Plot
#'
#' Draws a reaction norm or interaction plot as typically used in ecology to
#' demonstrate treatment x genotype interactions.  The x axis can be
#' categorical or numeric.  There are options to plot the raw data points in
#' addition to a choice of several summary methods.  The central values may be
#' connected or the entire data set fitted to a line.  A summary table is
#' optional, and gives either counts at each level, an ANOVA table or fit data.
#'
#' The choices in \code{method} are based upon functions in the
#' \code{\link[ChemoSpec]{ChemoSpec}} package which will need to be installed.
#' See \code{ChemoSpec::.seXy} for the details.
#'
#' @param formula A formula with both LHS and RHS.  The formula should comply
#'        with the usual \code{lattice} conventions, for instance \code{y ~ x}.
#'        The formula entries refer to named columns in \code{data}.
#'
#' @param data A data frame containing the response (y) values, categorical/factor
#'        values (x; numeric or factor), and group values (factor).
#'
#' @param groups The (unquoted) name of a column in \code{data} to serve as
#'        the grouping variable.  The grouping variable typically represents genotype,
#'        and the summary statistics are based on aggregating individuals of a given
#'        genotype.  However, if \code{type = "raw"} the grouping variable identifies
#'        unique individuals for which there are multiple values of the x variable.
#'        \code{groups} is required.
#'
#' @param cols A vector of colors to be used for the x variable given in the
#'        in the formula.  Must be as long as the levels in the variable.
#'
#' @param freckles Logical; if \code{TRUE}, the actual data points are plotted
#'        as small points, hence the name, jittered slightly in the horizontal
#'        direction only.  Ignored when \code{type = "raw"}.
#'
#' @param type Either "connect", "fitLine" , "anova", "anovaNoLine" or "raw".
#'        \itemize{
#'           \item {For "connect", a line is drawn which connects the chosen measure
#'                 of central tendency (e.g. mean) for each value of x.
#'                 \emph{Used when the x valuea are categorical.}}
#'           \item {For "anova", a connecting line is drawn along with an ANOVA summary table.
#'                 \emph{Used when the x values are categorical.}}
#'           \item {For "anovaNoLine", no line is drawn and the ANOVA summary table is provided.,
#'                 \emph{Used when the x values are categorical.}}
#'           \item {For "raw", each observation/sample is shown as a single point
#'                 (no summary statistics) and connected to its value at each level of
#'                 the categorial variable. \emph{Used when the x values are categorical.}}
#'           \item {For "fitLine", a linear fit is performed and
#'                 the line is drawn.  \emph{used for continuous x values.}}
#'        }
#'
#' @param method One of \code{c("sem", "sem95", "iqr", "mad")}.  Each produces
#'        a dot at the measure of central tendency and error bars corresponding to the
#'        requested method ("box" currently does not work quite right).  Full details
#'        are in \code{ChemoSpec::.seXy}.   Ignored when \code{type = "raw"} as no
#'        summary stats are done and each individual sample gets its own line between
#'        levels of the x variable.
#'
#' @param table If NULL, the default, no action.  If a vector of 3 numbers
#'        \code{c(x, y, cex)}, a summary table is placed at the given coordinates with
#'        font size \code{cex}.  The table contains either a summary of counts (if \code{type
#'        = "connect"}), a report of the linear fit (if \code{type = "fitLine"}) or an
#'        ANOVA table (if \code{type = "anova"}).  The coordinates are in device units
#'        so 0.5, 0.5 puts the table dead center in the plotting region.
#'        Ignored when \code{type = "raw"}.
#'
#' @param legend If NULL, no action.  If a vector of 3 numbers
#'         \code{c(x, y, cex)}, a legend is placed at the given coordinates with
#'         font size \code{cex}.  The coordinates are in device units
#'         so 0.5, 0.5 puts the legend dead center in the plotting region.
#'         Ignored when \code{type = "raw"}.
#'
#' @param theme Character; A suitable \code{lattice} theme.
#'        There are two built-in themes which you can use as is or modify to your heart's
#'        content.  If  none is given, \code{\link{screenTheme}} will be used.  The other option
#'        provided is \code{\link{posterTheme}}.
#'
#' @param \dots Additional arguments to be passed downstream to lattice functions.
#'
#' @return A \code{lattice} plot object is returned invisibly, which must be 
#'        "printed" to create a plot.
#'
#' @author Bryan A. Hanson, DePauw University. \email{hanson@@depauw.edu}
#' @references \url{https://github.com/bryanhanson/HandyStuff}
#' @keywords univariate plot
#' @export
#'
#' @examples
#' 
#' require("ChemoSpec")
#' require("lattice")
#' require("gridExtra")
#'
#' ### set up demo data
#' 
#' set.seed(79)
#' res <- c( # response values
#'   rnorm(10, 5, 1.5), rnorm(10, 8, 2),
#'   rnorm(10, 14, 2.0), rnorm(10, 10, 1.5),
#'   rnorm(10, 15, 2), rnorm(10, 12, 2.5)
#' )
#' fac1 <- c(rep("L", 20), rep("M", 20), rep("H", 20)) # treatment: categorical x values
#' fac1 <- factor(fac1, levels = c("L", "M", "H"))
#' fac2 <- sample(c("A", "B"), 60, replace = TRUE) # genotype, used for grouping summaries
#' fac2 <- as.factor(fac2)
#' num <- c(rep(5, 20), rep(10, 20), rep(12, 20)) # treatment: continuous x values
#' td1 <- data.frame(resp = res, fac1 = fac1, fac2 = fac2, num = num)
#' 
#' ### Numeric x with lm and summary table:
#' 
#' p <- rxnNorm(
#'   formula = resp ~ num, groups = fac2, data = td1,
#'   method = "iqr", freckles = TRUE, type = "fitLine",
#'   cols = c("red", "blue"), table = c(0.5, 0.3, 0.75),
#'   legend = c(0.175, 0.84, 1.0),
#'   main = "rxnNorm - linear model w/fit table"
#' )
#' print(p)
#' 
#' ### Categorical x, connected and summary table:
#' 
#' p <- rxnNorm(
#'   formula = resp ~ fac1, groups = fac2, data = td1,
#'   method = "iqr", freckles = TRUE, type = "connect",
#'   cols = c("red", "blue"), table = c(0.5, 0.3, 0.75),
#'   legend = c(0.175, 0.84, 1.0),
#'   main = "rxnNorm - Categorical x w/summary table"
#' )
#' print(p)
#' 
#' ### Categorical x, connected with ANOVA table:
#' #
#' p <- rxnNorm(
#'   formula = resp ~ fac1, groups = fac2, data = td1,
#'   method = "iqr", freckles = TRUE, type = "anova",
#'   cols = c("red", "blue"), table = c(0.57, 0.2, 0.75),
#'   legend = c(0.175, 0.84, 1.0),
#'   main = "rxnNorm - Categorical x w/ANOVA table"
#' )
#' print(p)
#'
#' ### Categorical x, plot each samples' raw values, no summaries or tables:
#'
#' ### set up demo data
#' 
#' res <- abs(rnorm(15)) # response values
#' fac1 <- rep(c("L", "M", "H"), times = 5) # treatment: categorical x values
#' fac1 <- factor(fac1, levels = c("L", "M", "H"))
#' fac2 <- c(rep("A", 3), rep("B", 3), rep("C", 3), rep("D", 3), rep("E", 3))
#' fac2 <- as.factor(fac2) # grouping variable, each sample/individual is a group
#' cols <- c(rep("black", 3), rep("red", 3), rep("blue", 3), rep("green", 3), rep("orange", 3))
#' td2 <- data.frame(resp = res, fac1 = fac1, fac2 = fac2, col = cols)
#'
#' p <- rxnNorm(
#'   formula = resp ~ fac1, groups = fac2, data = td2, cols = cols,
#'   type = "raw", main = "rxnNorm - Categorical x w/Raw Sample Values"
#' )
#' print(p)
#'
rxnNorm <-
  function(formula = NULL, data = NULL, groups = NULL,
           cols = NULL, freckles = FALSE, type = "connect",
           method = c("sem", "sem95", "iqr", "mad"),
           table = NULL, legend = c(0.5, 0.5, 1.0),
           theme = screenTheme(), ...) {

    # Function to compare categorical data by
    # plotting means etc in the same panel
    # Options to connect the dots or do a fit
    # and display some nice summary tables
    # Bryan Hanson, DePauw Univ, July 2010

    # Lattice version, February 2013
    # We are using formula, data and groups close to how lattice handles them

    if (!requireNamespace("lattice", quietly = TRUE)) {
      stop("You need to install package lattice to use this function")
    }
    if (!requireNamespace("latticeExtra", quietly = TRUE)) {
      stop("You need to install package latticeExtra to use this function")
    }
    if (!requireNamespace("plyr", quietly = TRUE)) {
      stop("You need to install package plyr to use this function")
    }
    if (!requireNamespace("RColorBrewer", quietly = TRUE)) {
      stop("You need to install package RColorBrewer to use this function")
    }
    if (!requireNamespace("ChemoSpec", quietly = TRUE)) {
      stop("You need to install package ChemoSpec to use this function")
    }
    if (!requireNamespace("gridExtra", quietly = TRUE)) {
      stop("You need to install package gridExtra to use this function")
    }
    if (!requireNamespace("gtable", quietly = TRUE)) {
      stop("You need to install package gtable to use this function")
    }
    if (!requireNamespace("grid", quietly = TRUE)) {
      stop("You need to install package grid to use this function")
    }

    # Check the input and process the formula & data types

    if (is.null(formula)) stop("Formula not given")
    if (!plyr::is.formula(formula)) stop("Invalid formula specification")

    yy <- as.character(formula[[2]])
    if (!length(formula[[3]]) == 1) stop("Formula must be of type y ~ x")
    if (length(formula[[3]]) == 1) {
      xx <- as.character(formula[[3]])
    }

    if (is.null(groups)) stop("You must provide groups as a factor")
    grn <- as.character(match.call()$groups) # the name
    if (!is.factor(data[, grn])) stop("Groups must be a factor")

    if (type == "fitLine") {
      if (!is.numeric(data[, xx])) {
        msg <- paste(xx, "must be numeric to fit a line")
        stop(msg)
      }
    }

    # Reduce the df and clean it of NAs
    # otherwise there is trouble with the counts

    keep <- c(yy, xx, grn)
    data <- data[, keep]
    data <- stats::na.omit(data)

    # Set up the color scheme if not provided

    nc <- length(levels(data[, grn]))
    if (is.null(cols)) {
      cols <- RColorBrewer::brewer.pal(nc, "Set1")
      if (nc == 2) cols <- cols[1:2] # cannot keep the extra due to recycling problems
    }
    if (!is.null(cols)) {
      if (type != "raw") {
      if (!length(cols) == nc) stop("cols must be the same length as the levels in groups")
}
    }

    args <- as.list(match.call(expand.dots = FALSE)[-1]) # used a bit later

    ### This panel will compute & plot summary statistics

    panel.summary <- function(x, y, ...) {
      mean <- stats::aggregate(data[, yy] ~ data[, xx] * data[, grn], data, FUN = mean)
      med <- stats::aggregate(data[, yy] ~ data[, xx] * data[, grn], data, FUN = stats::median)
      sexy <- stats::aggregate(data[, yy] ~ data[, xx] * data[, grn], data, FUN = ChemoSpec::.seXy)
      sexy95 <- stats::aggregate(data[, yy] ~ data[, xx] * data[, grn], data, FUN = ChemoSpec::.seXy95)
      sexymad <- stats::aggregate(data[, yy] ~ data[, xx] * data[, grn], data, FUN = ChemoSpec::.seXyMad)
      sexyiqr <- stats::aggregate(data[, yy] ~ data[, xx] * data[, grn], data, FUN = ChemoSpec::.seXyIqr)
      sumDat <- cbind(
        mean, med[, 3], sexy[[3]][, c(2, 3)],
        sexy95[[3]][, c(2, 3)], sexymad[[3]][, c(2, 3)], sexyiqr[[3]][, c(2, 3)]
      )
      names(sumDat) <- c(
        "xx", "gr", "mean", "median",
        "semL", "semU", "sem95L", "sem95U", "madL", "madU",
        "iqrL", "iqrU"
      )
      gr <- NULL # Needed for next command to fake out check mechanism
      sumDat <- plyr::arrange(sumDat, xx, gr)
      sumDat$cols <- rep(cols, length(unique(sumDat$xx)))

      for (i in 1:nrow(sumDat)) {

        # Plot the mean or median

        if ((method == "sem") | (method == "sem95")) {
          lattice::panel.points(x = sumDat[i, 1], y = sumDat[i, 3], col = sumDat[i, 13], pch = 19, ...)
        }

        if ((method == "mad") | (method == "iqr")) {
          lattice::panel.points(x = sumDat[i, 1], y = sumDat[i, 4], col = sumDat[i, 13], pch = 19, ...)
        }

        # Plot the 'whiskers'

        if (method == "sem") {
          lattice::panel.segments(
            x0 = sumDat[i, 1], y0 = sumDat[i, 5],
            x1 = sumDat[i, 1], y1 = sumDat[i, 6], lwd = 3, col = sumDat[i, 13], ...
          )
        }

        if (method == "sem95") {
          lattice::panel.segments(
            x0 = sumDat[i, 1], y0 = sumDat[i, 7],
            x1 = sumDat[i, 1], y1 = sumDat[i, 8], lwd = 3, col = sumDat[i, 13], ...
          )
        }

        if (method == "mad") {
          lattice::panel.segments(
            x0 = sumDat[i, 1], y0 = sumDat[i, 9],
            x1 = sumDat[i, 1], y1 = sumDat[i, 10], lwd = 3, col = sumDat[i, 13], ...
          )
        }

        if (method == "iqr") {
          lattice::panel.segments(
            x0 = sumDat[i, 1], y0 = sumDat[i, 11],
            x1 = sumDat[i, 1], y1 = sumDat[i, 12], lwd = 3, col = sumDat[i, 13], ...
          )
        }
      }
    } # end of panel.summary

    ### This panel will connect the means or medians

    panel.connect <- function(x, y, ...) {
      mean <- stats::aggregate(data[, yy] ~ data[, xx] * data[, grn], data, FUN = mean)
      med <- stats::aggregate(data[, yy] ~ data[, xx] * data[, grn], data, FUN = stats::median)
      sumDat <- cbind(mean, med[, 3])
      names(sumDat) <- c("xx", "gr", "mean", "median")
      gr <- NULL # needed to fake out check mechanism for next command
      sumDat <- plyr::arrange(sumDat, gr, xx)
      sumDat$cols <- rep(cols, each = length(unique(sumDat$xx)))

      # The pattern here took a long time to come up with!
      # The premise is that for a nlvls x ngr plot, you have to
      # connect only certain pairs of points, in particular
      # every nlvls_th point pair has to be skipped to
      # create a universal pattern

      nlvls <- length(unique(sumDat$xx))
      ngr <- length(unique(sumDat$gr))

      i1 <- 1:(nlvls * ngr)
      rem <- seq(nlvls, (nlvls * ngr), nlvls)
      i1 <- i1[-rem]

      # Troubleshooting commented out (but keep!)

      # 		print(sumDat)
      # 		cat("no. levels =", nlvls, "\n")
      # 		cat("no. groups =", ngr, "\n")
      # 		cat("Rows from sumDat are connected in vertical pairs as follows\n"s)
      # 		cat("first point = ", i1, "\n")
      # 		i2 <- i1 + 1
      # 		cat("second point = ", i2, "\n")

      for (i in i1) {
        if ((method == "sem") | (method == "sem95")) {
          lattice::panel.segments(
            x0 = sumDat[i, 1], y0 = sumDat[i, 3],
            x1 = sumDat[i + 1, 1], y1 = sumDat[i + 1, 3], col = sumDat[i, 5], ...
          )
        }
        if ((method == "mad") | (method == "iqr")) {
          lattice::panel.segments(
            x0 = sumDat[i, 1], y0 = sumDat[i, 4],
            x1 = sumDat[i + 1, 1], y1 = sumDat[i + 1, 4], col = sumDat[i, 5], ...
          )
        }
      }
    } # end of panel.connect

    ### This panel will connect the individual samples across values of a categorical
    ### variable when type = "raw"

    panel.connect2 <- function(x, y, ...) {
      data <- cbind(data, cols)
      f2lvls <- levels(data$fac2)
      nf2 <- length(f2lvls)
      nf1 <- length(levels(data$fac1))
      for (i in 1:nf2) {
        tmp <- data[data$fac2 == f2lvls[i], ]
        for (j in 1:(nf1 - 1)) {
          lattice::panel.segments(
            x0 = tmp[j, "fac1"],
            x1 = tmp[j + 1, "fac1"],
            y0 = tmp[j, "resp"],
            y1 = tmp[j + 1, "resp"],
            col = tmp[j, "cols"]
          )
        }
      }
    } # end of panel.connect2

    ### Now prepare summary tables using tableGrob from gridExtra

    if (!is.null(table)) {
      TT <- gridExtra::ttheme_default(
        core = list(fg_params = list(cex = table[3])),
        colhead = list(fg_params = list(cex = table[3])),
        rowhead = list(fg_params = list(cex = table[3]))
      )

      if (type == "connect") {
        if (!is.factor(data[, xx])) {
          msg <- paste(xx, "must be a factor for a table of counts")
          stop(msg)
        }
        counts <- plyr::count(data, vars = c(grn, xx))
        colnames(counts) <- c(grn, xx, "count")

        myt <- gridExtra::tableGrob(counts, rows = NULL, theme = TT)
        myt <- gtable::gtable_add_grob(myt,
          grobs = grid::rectGrob(gp = grid::gpar(fill = NA, lwd = 2)),
          t = 2, b = nrow(myt), l = 1, r = ncol(myt)
        )
        myt <- gtable::gtable_add_grob(myt,
          grobs = grid::rectGrob(gp = grid::gpar(fill = NA, lwd = 2)),
          t = 1, l = 1, r = ncol(myt)
        )
      }

      if (type == "anova") {
        if (!is.factor(data[, xx])) {
          msg <- paste(xx, "must be a factor for anova")
          stop(msg)
        }
        f <- paste(".~.*", deparse(substitute(groups)))
        nf <- stats::update.formula(formula, f)
        mod <- stats::aov(formula = nf, data = data)

        myt <- prettyAOVtable(aov.object = mod, cex = table[3])
      }

      if (type == "fitLine") {
        lvls <- levels(data[, grn])
        nl <- length(lvls)
        m <- c()
        b <- c()
        r2 <- c()
        for (i in 1:nl) {
          wh <- grepl(lvls[i], data[, grn])
          dat <- data[wh, ]
          mod <- stats::lm(dat[, yy] ~ dat[, xx])
          m[i] <- round(mod$coef[2], 2)
          b[i] <- round(mod$coef[1], 2)
          r2[i] <- round(stats::cor(dat[, xx], dat[, yy])^2, 4)
        }

        mod.yy <- data.frame(line = lvls, m = m, b = b, r2 = r2)

        myt <- gridExtra::tableGrob(mod.yy, rows = NULL, theme = TT)
        myt <- gtable::gtable_add_grob(myt,
          grobs = grid::rectGrob(gp = grid::gpar(fill = NA, lwd = 2)),
          t = 2, b = nrow(myt), l = 1, r = ncol(myt)
        )
        myt <- gtable::gtable_add_grob(myt,
          grobs = grid::rectGrob(gp = grid::gpar(fill = NA, lwd = 2)),
          t = 1, l = 1, r = ncol(myt)
        )
      }
    } # End of table preparations

    ##### Now the high level plot calls

    ps <- theme

    p <- lattice::xyplot(stats::as.formula(args$formula),
      data = eval(args$data),
      groups = eval(args$groups), ...,
      axis = latticeExtra::axis.grid,
      par.settings = ps,
      panel = function(x, y, ...) {
        if (type != "raw") {
          panel.summary(x, y, ...)
          if (freckles) lattice::panel.xyplot(x, y, jitter.x = TRUE, col = cols, ...)
          if ((type == "connect") | (type == "anova")) panel.connect(x, y, ...)
          if (type == "fitLine") lattice::panel.xyplot(x, y, type = "r", col = cols, ...)
          if (!is.null(table)) {
            if (type == "connect") {
              grid::upViewport(0)
              tv <- grid::viewport(x = table[1], y = table[2])
              grid::pushViewport(tv)
              grid::grid.draw(myt)
            }
            if (type == "anova") {
              grid::upViewport(0)
              tv <- grid::viewport(x = table[1], y = table[2])
              grid::pushViewport(tv)
              grid::grid.draw(myt)
            }
            if (type == "fitLine") {
              grid::upViewport(0)
              tv <- grid::viewport(x = table[1], y = table[2])
              grid::pushViewport(tv)
              grid::grid.draw(myt)
            }
          } # end of table options

          if (!is.null(legend)) {
            TTT <- gridExtra::ttheme_minimal(core = list(fg_params = list(cex = legend[3], col = cols)))

            legtxt <- data.frame(lvs = levels(data[, grn]))
            myleg <- gridExtra::tableGrob(legtxt,
              cols = NULL, rows = NULL, theme = TTT
            )
            myleg <- gtable::gtable_add_grob(myleg,
              grobs = grid::rectGrob(gp = grid::gpar(fill = NA, lwd = 2)),
              t = 1, b = nrow(myleg), l = 1, r = ncol(myleg)
            )
            grid::upViewport(0)
            lv <- grid::viewport(x = legend[1], y = legend[2])
            grid::pushViewport(lv)
            grid::grid.draw(myleg)
          } # end of legend processing
        } # end of type != "raw

        if (type == "raw") {
          panel.connect2(x, y, ...)
        }
      } # end of panel function
    ) # end of xyplot

    invisible(p)
  }
